packages/flutter/lib/src/material/navigation_bar.dart

```diff
@@ -0,0 +1,1144 @@
+      color: backgroundColor ?? ElevationOverlay.colorWithOverlay(colorScheme.surface, colorScheme.onSurface, 3.0),
+      child: NavigationBarBottomPadding(
+        child: SizedBox(
+          height: effectiveHeight,
+          child: Row(
+            children: <Widget>[
+              for (int i = 0; i < destinations.length; i++)
+                Expanded(
+                  child: SelectableAnimatedBuilder(
+                    duration: animationDuration,
+                    isSelected: i == selectedIndex,
+                    builder: (BuildContext context, Animation<double> animation) {
+                      return NavigationBarDestinationInfo(
+                        destinationNumber: i,
+                        totalNumberOfDestinations: destinations.length,
+                        selectedAnimation: animation,
+                        labelBehavior: labelBehavior,
+                        onTap: onDestinationSelected != null ? () => onDestinationSelected!(i) : () {},
+                        child: destinations[i],
+                      );
+                    },
+                  ),
+                ),
+            ],
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+/// Options for how the labels should layout and when they should appear in the
+/// navigation bar destinations.
+enum NavigationBarDestinationLabelBehavior {
+  /// Always shows all of the labels under each navigation bar destination,
+  /// selected and unselected.
+  alwaysShow,
+
+  /// Never shows any of the labels under the navigation bar destinations,
+  /// regardless of selected vs unselected.
+  alwaysHide,
+
+  /// Only shows the labels of the selected navigation bar destination.
+  ///
+  /// When an destination is unselected, the label will be faded out, and the
+  /// icon will be centered.
+  ///
+  /// When an destination is selected, the label will fade in and the label and
+  /// icon will slide up so that they are both centered.
+  onlyShowSelected,
+}
+
+/// Destination Widget for displaying Icons + labels in the Material 3
+/// Navigation Bars through [NavigationBar.destinations].
+///
+/// The destination this widget creates will look something like this:
+/// =======
+/// |
+/// |  ☆  <-- [icon] (or [selectedIcon])
+/// | text <-- [label]
+/// |
+/// =======
+///
+/// See also:
+///  * [NavigationBarDestinationBuilder] - Use this class when you need more
+///    custom destinations than [NavigationBarDestination] provides, for
+///    example, if you have animated icons.
+class NavigationBarDestination extends StatelessWidget {
+  /// Creates a navigation bar destination with an icon and a label, to be used
+  /// in the [NavigationBar.destinations].
+  const NavigationBarDestination({
+    Key? key,
+    required this.icon,
+    this.selectedIcon,
+    required this.label,
+    this.tooltip,
+  }) : super(key: key);
+
+  /// The [Widget] (usually an [Icon]) that displays when this
+  /// [NavigationBarDestination] is selected.
```

Comment:
The documentation for this and `selectedIcon` should be partially swapped (I think you used to have them the other way)